homeassistant/components/sensor/nextbus.py

```diff
@@ -0,0 +1,271 @@
+
+        return self._name
+
+    @property
+    def state(self):
+        """Return current state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return ATTR_MINUTES
+
+    @property
+    def device_state_attributes(self):
+        """Return additional state attributes."""
+        return self._attributes
+
+    @property
+    def icon(self):
+        """Return icon to be used for this sensor."""
+        # Would be nice if we could determine if the line is a train or bus
+        # however that doesn't seem to be available to us. Using bus for now.
+        return ICON
+
+    def update(self):
+        """Update sensor with new departures times."""
+        # Note: using Multi because there is a bug with the single stop impl
+        predictions = get_predictions(
+            [{
+                'stop_tag': int(self.stop),
+                'route_tag': self.route,
+            }],
+            self.agency,
+        )
+
+        self._logger.debug('Predictions results: %s', predictions)
+
+        if 'Error' in predictions:
+            self._logger.error('Could not get predictions: %s', predictions)
+
+        if not predictions.get('predictions'):
+            self._logger.info('No predictions available')
+            self._state = None
+            # Remove attributes that may now be outdated
+            self._attributes.pop('Upcoming', None)
+            return
+
+        predictions = predictions['predictions']
+
+        # Set detailed attributes
+        self._attributes.update({
+            'Agency': predictions.get('agencyTitle'),
+            'Route': predictions.get('routeTitle'),
+            'Direction': predictions.get('direction', {}).get('title'),
+            'Stop': predictions.get('stopTitle'),
+        })
+
+        # Sometimes message will return a dict, others a list of dicts
+        message = predictions.get('message')
+        if isinstance(message, dict):
+            self._attributes['Message'] = message.get('text', '')
+        elif isinstance(message, list):
+            self._attributes['Message'] = ' -- '.join(
+                m.get('text', '')
+                for m in message
+            )
+        else:
+            self._attributes['Message'] = ''
+
+        # Maybe offer some custom formatting here
+        self._name = '{Agency} - {Route}'.format(
+            **self._attributes
+        )
+
+        upcoming = predictions.get('direction', {}).get('prediction')
+        if not upcoming:
+            self._logger.info('No upcoming predictions available')
+            self._state = None
+            self._attributes['Upcoming'] = 'No upcoming predictions'
+            return
+
+        self._attributes['Upcoming'] = ', '.join(
+            p['minutes'] for p in upcoming
+        )
+
+        latest_prediction = upcoming[0]
+        self._state = latest_prediction['minutes']
```

Comment:
Instead of using minutes to next departure as state, convert that to an absolute utc timestamp by adding the timedelta minutes to `utcnow` from our dt.py util.

Then set the sensor `device_class` property to `timestamp`. This will allow the user to select a relative time representation in the frontend and we avoid not needed state changes only caused by a time change.

https://developers.home-assistant.io/docs/en/entity_sensor.html#available-device-classes

Reply:
>[...] we avoid not needed state changes only caused by a time change.

What do you mean by this? Will it not call update as frequently if a timestamp is returned?

Also, is there an example of this? Should state return the unix timestamp or a string? The table there says `ISO8601`.

Reply:
It will call update as before. The difference is that a new state won't be written carelessly just because time has passed since last update. Only if the timestamp changes will a new state be written.

The state attributes can also cause a state change though, so we won't avoid this problem totally. If possible avoid relative time points there too.

Reply:
See eg:
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/gtt/sensor.py#L82-L83